### PR TITLE
Dropping environment variables

### PR DESCRIPTION
**Pull request by matajoh**
Putting this here for initial review. The fundamental change here is that `EnforceCreateContainerPolicy`, `EnforceExecInContainerPolicy`, and `EnforceExecExternalProcessPolicy` now (optionally) return an updated environment variable list if `policy.DropEnvironmentVariables` is set to true. This requires some new logic when performing narrowing:

``` rego
valid_envs_subset(env_rules) := envs {
    envs := {env |
        some env in input.envList
        some rule in env_rules
        env_ok(rule.pattern, rule.strategy, env)
    }
}

valid_envs_for_all(items) := envs {
    data.policy.drop_environment_variables

    # for each item, find a subset of the environment rules
    # that are valid
    valid := [envs |
        some item in items
        envs := valid_envs_subset(item.env_rules)
    ]
    
    # we want to select the most specific matches, which in this
    # case consists of those matches which require dropping the
    # fewest environment variables (i.e. the longest lists)
    counts := [num_envs |
        envs := valid[_]
        num_envs := count(envs)
    ]
    max_count := max(counts)

    largest_env_sets := {envs |
        some i
        counts[i] == max_count
        envs := valid[i]
    }

    # if there is more than one set with the same size, we
    # can only proceed if they are all the same, so we verify
    # that the intersection is equal to the union. For a single
    # set this is trivially true.
    envs_i := intersection(largest_env_sets)
    envs_u := union(largest_env_sets)
    envs_i == envs_u
    envs := envs_i
}

valid_envs_for_all(items) := envs {
    not data.policy.drop_environment_variables
    # no dropping allowed, so we just return the input
    envs := input.envList
}
```

here is an example of container narrowing using this logic:

```
    # narrow the matches based upon command, working directory, and
    # mount list
    possible_containers := [container |
        some container in data.metadata.matches[input.containerID]
        workingDirectory_ok(container.working_dir)
        command_ok(container.command)
        mountList_ok(container.mounts, container.allow_elevated)
    ]

    count(possible_containers) > 0

    # check to see if the environment variables match, dropping
    # them if allowed (and necessary)

    env_list := valid_envs_for_all(possible_containers)
    containers := [container |
        some container in possible_containers
        envList_ok(container.env_rules, env_list)
    ]
```

### CODE REVIEW

File: internal/guest/runtime/hcsv2/uvm.go [hunk 1]
```diff
@@ -318,6 +319,10 @@ func (h *Host) CreateContainer(ctx context.Context, id string, settings *prot.VM
 		return nil, errors.Wrapf(err, "container creation denied due to policy")
 	}
 
+	if envToKeep != nil {
+		settings.OCISpecification.Process.Env = []string(envToKeep)
```
Comment (matajoh) [inline internal/guest/runtime/hcsv2/uvm.go:323]: @anmaxvl Is this (and the other two changes in this file) the right way to do this environment variable dropping? What is the best way to test this more fully?

File: pkg/securitypolicy/framework.rego [hunk 2]
```diff
@@ -244,16 +306,29 @@ mountList_ok(mounts, allow_elevated) {
...
         workingDirectory_ok(container.working_dir)
         some process in container.exec_processes
         command_ok(process.command)
     ]
+
+    count(possible_containers) > 0
+
+    # check to see if the environment variables match, dropping
+    # them if allowed (and necessary)
+
```
Comment (SeanTAllen) [inline pkg/securitypolicy/framework.rego:323]: see above

File: pkg/securitypolicy/framework.rego [hunk 3]
```diff
@@ -355,24 +430,41 @@ enforcement_point_info := {"available": false, "allowed": false, "unknown": fals
...
+        some process in feed.external_processes
+        workingDirectory_ok(process.working_dir)
+        command_ok(process.command)
+    ]
+
+    possible_processes := array.concat(policy_processes, fragment_processes)
+
+    # check to see if the environment variables match, dropping
+    # them if allowed (and necessary)
+
```
Comment (SeanTAllen) [inline pkg/securitypolicy/framework.rego:460]: see above.

File: pkg/securitypolicy/regopolicy_test.go [hunk 4]
```diff
@@ -601,6 +602,192 @@ func Test_Rego_EnforceEnvironmentVariablePolicy_NotAllMatches(t *testing.T) {
...
+		actual, err := tc.policy.EnforceCreateContainerPolicy(tc.sandboxID, tc.containerID, tc.argList, envList, tc.workingDir, tc.mounts)
+
+		// getting an error means something is broken
+		if err != nil {
+			t.Errorf("Expected container creation to be allowed. It wasn't: %v", err)
+			return false
+		}
+
+		if !areStringArraysEqual(actual, tc.envList) {
+			t.Errorf("environment variables were not dropped correctly.")
```
Comment (SeanTAllen) [inline pkg/securitypolicy/regopolicy_test.go:629]: this message seems misleading. in this test, no environment variables are dropped, right?

File: pkg/securitypolicy/securitypolicy_internal.go [hunk 5]
```diff
@@ -44,32 +45,32 @@ func newSecurityPolicyInternal(containers []*Container,
...
-	Signals []syscall.Signal
+	Signals []syscall.Signal `json:"signals"`
 }
 
 type containerExecProcess struct {
-	Command []string
+	Command []string `json:"command"`
 	// A list of signals that are allowed to be sent to this process
-	Signals []syscall.Signal
+	Signals []syscall.Signal `json:"signals"`
```
Comment (SeanTAllen) [inline pkg/securitypolicy/securitypolicy_internal.go:73]: see json annotation added question

File: pkg/securitypolicy/securitypolicy_internal.go [hunk 6]
```diff
@@ -80,10 +81,10 @@ type externalProcess struct {
...
 // Internal version of Mount
 type mountInternal struct {
-	Source      string
-	Destination string
-	Type        string
-	Options     []string
+	Source      string   `json:"source"`
+	Destination string   `json:"destination"`
+	Type        string   `json:"type"`
+	Options     []string `json:"options"`
```
Comment (SeanTAllen) [inline pkg/securitypolicy/securitypolicy_internal.go:87]: see json annotation added question

File: pkg/securitypolicy/securitypolicy_test.go [hunk 7]
```diff
@@ -27,17 +27,17 @@ const (
...
+	maxGeneratedEnvironmentVariableRules      = 8
 	maxGeneratedMountTargetLength             = 256
 	rootHashLength                            = 64
 	maxGeneratedMounts                        = 4
 	maxGeneratedMountSourceLength             = 32
 	maxGeneratedMountDestinationLength        = 32
 	maxGeneratedMountOptions                  = 5
 	maxGeneratedMountOptionLength             = 32
-	maxGeneratedExecProcesses                 = 12
+	maxGeneratedExecProcesses                 = 4
```
Comment (SeanTAllen) [inline pkg/securitypolicy/securitypolicy_test.go:40]: why the changes in maxes?

**Comment by SeanTAllen** [comment]
I believe this is missing code to make the generated content coming from the securitypolicy tool correctly recognize the config value and write it into the rego.

File: pkg/securitypolicy/securitypolicy_internal.go [hunk 8]
```diff
@@ -44,32 +45,32 @@ func newSecurityPolicyInternal(containers []*Container,
 // Internal version of Container
 type securityPolicyContainer struct {
 	// The command that we will allow the container to execute
-	Command []string
+	Command []string `json:"command"`
```
[the remark this answers is not in the document]
Comment (matajoh) [reply pkg/securitypolicy/securitypolicy_internal.go:48]: I wanted to do a deep copy of securityPolicy (for the tests) and the JSON serialization/deserialization trick was the easiest way to do it in a maintainable way.

File: pkg/securitypolicy/securitypolicy_test.go (hunk 7 above)
Comment (matajoh) [reply pkg/securitypolicy/securitypolicy_test.go:40]: because of the new environment variables logic individual tests for exec process in container and a few others were timing out after 30 seconds. Reducing the max values fixed the issue.
